### DISCUSSION TITLE
AWS amazon linux 2 Unable to load shared library Magick.Native-Q16-arm64.dll

### DISCUSSION THREAD

**Discussion by Flop007** [dlemstra/Magick.NET#1432] [Help]
Hello!,
I use AWS with EC2 amazon linux 2 and NGINX.
My project use .Net 6.0 and is build in Azure DevOps and send by himself at the EC2.

I always same Error :
System.DllNotFoundException: Unable to load shared library 'Magick.Native-Q16-arm64.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libMagick.Native-Q16-arm64.dll: cannot open shared object file: No such file or directory.

I add in my project [https://github.com/dlemstra/Magick.NET/blob/main/docs/CrossPlatform.md](url)
### <Code> .csproj
    <PropertyGroup>
        <MagickCopyNativeWindows>true</MagickCopyNativeWindows>
        <MagickCopyNativeLinux>true</MagickCopyNativeLinux>
        <MagickCopyNativeLinuxMusl>true</MagickCopyNativeLinuxMusl>
        <MagickCopyNativeMacOS>true</MagickCopyNativeMacOS>
    </PropertyGroup>
</Code>

Same problem

I add 
### <code> .csproj
	<Target Name="CopyRequiredNativeAssets" AfterTargets="_GenerateFunctionsPostBuild">
		<ItemGroup>
			<NativeAssetToCopy Include="$(OutDir)\Magick.Native-*.so" />
		</ItemGroup>
		<Copy SourceFiles="@(NativeAssetToCopy)" DestinationFolder="$(OutDir)\bin" />
	</Target>
</code>

Same

On the server EC2 amazon linux 2 in root project there are dependencies Magick.Native.
![image](https://github.com/dlemstra/Magick.NET/assets/143707954/34ee4f00-8997-4f0c-ae18-60c6cd90eedc)

I couldn't find a solution on the internet, can you help me?

**Comment by dlemstra** [comment]
You are probably missing some kind of dependency for that library. You might get a bit more information with `ldd libMagick.Native-Q16-arm64.dll.so`

**Comment by Flop007** [comment]
there is result
ldd: warning: you do not have execution permission for `./Magick.Native-Q16-arm64.dll.so'
./Magick.Native-Q16-arm64.dll.so: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by ./Magick.Native-Q16-arm64.dll.so)
        linux-vdso.so.1 (0x0000ffff8ba2d000)
        libresolv.so.2 => /lib64/libresolv.so.2 (0x0000ffff89b8f000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x0000ffff89b5a000)
        libm.so.6 => /lib64/libm.so.6 (0x0000ffff89a99000)
        libc.so.6 => /lib64/libc.so.6 (0x0000ffff89913000)
        /lib/ld-linux-aarch64.so.1 (0x0000ffff8b9ef000)

And my version LDD
ldd (GNU libc) 2.26
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

The version Amazon Linux 2 use old version of GNU libc

I must downgrade Magick.NET who use GNU libc 2.26

**Comment by Flop007** [comment]
If you want to use GLIBC_2.27
Connect to the server in SSL and run the commands

yum -y groupinstall "Development Tools"
yum install nss_nis
yum install libffi-devel
mkdir glibc
cd glibc/
wget --no-check-certificate https://ftp.gnu.org/gnu/glibc/glibc-2.27.tar.gz
tar -xvf glibc-2.27.tar.gz
mkdir build
cd build
../configure --prefix=/usr --disable-profile --enable-add-ons --with-headers=/usr/include --with-binutils=/usr/bin
make && make install

**Comment by dlemstra** [comment]
Is there a docker image that I could check this with? For the `x64` build the image `amazon/aws-lambda-dotnet:6` is used to check the `ldd` status and patches where made to support earlier version of GLIBC. Not sure if this is possible but I could try to make it compatible with the image that you are using?

**Comment by dlemstra** [comment]
I have some local changes ready to resolve this so this should work in the next release but I would still like to have the name of an image I can use to check this.

**Comment by Flop007** [comment]
Thanks for your help.
I don't have a docker image to provide you.
but when you release the version I will test if it works.
Soon we will move to Kubernetes with Docker.
Amazon Linux 2 is an OS dating from the end of 2018 with a GLIBC 2.26 from 2017-08-02 (really old in computing)

**Comment by dlemstra** [comment]
The max version the next release will use is GLIBC 2.17 and I will check for myself if I can find an image that I can use to reproduce this.

**Comment by JulianDeclercq** [comment]
Hello, I encountered a similar issue about two months ago, and this thread was the top Google result when searching for a solution. I hope my experience can help someone facing the same problem.

My exceptions looked like this:

```
System.TypeInitializationException: The type initializer for 'NativeMagickSettings' threw an exception.
 ---> System.DllNotFoundException: Unable to load shared library 'Magick.Native-Q16-x64.dll' or one of its dependencies.
 Error loading shared library libresolv.so.2: No such file or directory (needed by /app/Magick.Native-Q16-x64.dll.so)
```

Magick.NET worked perfectly fine locally, but not when deploying to stage/production.

I made a sandbox project to figure out the problem and solution [here](https://github.com/JulianDeclercq/Magick.NET-Sandbox).

Adding `libc6-compat` to the Alpine system provided the missing `libresolv.so.2` library that Magick.NET needed. This package emulates the environment expected by applications developed for glibc-based systems, allowing them to run on musl libc-based Alpine without modification. Package size is **1.65kb**.

To add `libc6-compat` and fix the issue, add the following to the Dockerfile:

```
RUN apk add --no-cache libc6-compat
```
